Add Py2x flag

nanjekyejoannah opened the pull request:
This PR adds the Py2x flag:

```
Python 3.12.0a0 (heads/migration-dirty:5842fee697, May 23 2022, 00:25:04) [Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.flags
sys.flags(debug=0, inspect=0, interactive=0, optimize=0, dont_write_bytecode=0, no_user_site=0, no_site=0, ignore_environment=0, verbose=0, bytes_warning=0, py2x_warning=1, quiet=0, hash_randomization=1, isolated=0, dev_mode=False, utf8_mode=0, warn_default_encoding=0)
>>> 
```

ltratt commented on Doc/c-api/exceptions.rst:
```diff
@@ -1091,6 +1092,8 @@ the variables:
 +------------------------------------------+---------------------------------+----------+
 | :c:data:`PyExc_BytesWarning`             | :exc:`BytesWarning`             |          |
 +------------------------------------------+---------------------------------+----------+
+| :c:data:`PyExc_Py2xWarning`             | :exc:`Py2xWarning`             |          |
```
I think the alignment of this row has gone a bit off and a bit more whitespace needs to be added?

ltratt commented on Lib/subprocess.py:
```diff
@@ -298,6 +298,7 @@ def _args_from_interpreter_flags():
         'verbose': 'v',
         'bytes_warning': 'b',
         'quiet': 'q',
+        #'py2x_warning': '2',
```
Intentionally commented out?

nanjekyejoannah replied:
Thanks, it has no effect but I may have to just remove it rather than comment.

ltratt replied:
Unless there's a good reason for commented code, I prefer to delete it (commented code is a sort-of phobia of mine!).

vext01 commented on Doc/library/exceptions.rst:
```diff
@@ -855,6 +855,10 @@ The following exceptions are used as warning categories; see the
 
+.. exception:: Py2xWarning
+
+   Base class for warnings related to 3.x compatibility.
```
Should it be "related to 2.x compatibility"?

nanjekyejoannah replied:
Seems that we are warning when one write 2.x code, to signal for 3.x compatibility.

vext01 replied:
So it's correct? If so, OK :)

vext01 commented on Lib/test/test_cmd_line.py:
```diff
@@ -684,7 +684,7 @@ def test_xdev(self):
         code = ("import warnings; "
                 "print(' '.join('%s::%s' % (f[0], f[2].__name__) "
-                                "for f in warnings.filters))")
+                                "for f in warnings.filters))") 
```
Is this change intentional?

nanjekyejoannah replied:
Not, I will use `make patchcheck` to remove these spaces.

vext01 commented:
Just a couple of comments from me.

I'm surprised how much boilerplate is required, but it looks correct to me.

nanjekyejoannah commented:
Great, thanks, If its in good shape, I will just add some tests for `_embed` and the `sys` module, then I will use the flag for the `cmp` porting (In a second PR).

ltratt commented:
@nanjekyejoannah Works for me!

nanjekyejoannah commented:
I have addressed the changes and fixed the flag bug. Multiprocessing tests were exhibiting random behaviours, so I would appreciate, a review on that.

ltratt commented:
This is OK with me. @vext01 ?

vext01 commented on Lib/test/_test_multiprocessing.py:
```diff
@@ -3965,18 +3965,18 @@ def test_shared_memory_recreate(self):
 
             mock_make_filename.side_effect = names
-            shm2 = shared_memory.SharedMemory(create=True, size=1)
+            shm2 = shared_memory.SharedMemory(create=True, size=512)
```
we've changed some filenames and parameters here. Is that needed?

vext01 replied:
To elaborate, I don't see why adding a py2x flag would require these path and size changes. Maybe I missed something.

nanjekyejoannah replied:
I increased from 1 to 512, this test was failing so much for recreation cases. That is why I need some verification.

vext01 replied:
Was it failing before adding the flag?

nanjekyejoannah replied:
Not, but sometimes it gets skipped, so its random, it can run or not run, if you run the whole test suite. Anyway, I can revert it, no big deal.

vext01 replied:
Where did you get the idea to rename the files and up the size? Is that something you saw in the wild?

nanjekyejoannah replied:
Yes, the recreate test, would fail at times, the renaming, might have come with other changes that fixed it though. A particular conditional I removed in a config setup.

ltratt commented:
Should I do a bors try?

nanjekyejoannah commented:
Yes

ltratt commented:
bors try

ltratt commented:
I think we're ready to squash?

nanjekyejoannah commented:
Sure, let me do now

nanjekyejoannah committed:
Add a -2 flag to issue a warning related to 3.x compatibility

This is the equivalent of the -3 flag in 2.x. As with that flag, the aim is to ease Python 2->3 migration, though in this case in the opposite direction to the -3 flag

ltratt commented:
Could we add a bit of motivation to the commit message, even if only a couple of sentences? Please feel free to force push an update.

nanjekyejoannah commented:
Does this "Add a py2x_warning flag to issue a warning related to 3.x compatibility" look fine?

ltratt commented:
I am a fan of motivation (i.e. "why" rather than "what" since the latter is implicitly contained in the diff but the former is often hard to divine), so maybe adding a bit more would be good? e.g. "This is the equivalent of the `-2` switch in python2. As with that switch, the aim is to ease Python 2->3 migration, though in this case in the opposite direction to the `-2` switch." I'm sure you can do better than that, but hopefully it's an idea :)

nanjekyejoannah commented:
I made some edits to your wording in addition to the former "what" message.

ltratt commented:
We're almost there :)

We just need to add a double newline after "Add a -2 flag to issue a warning related to 3.x compatibility." otherwise git log will render everything terribly. [Commit messages are "Short description\n\nLong description".] Please force push that and then I'll merge :)

nanjekyejoannah commented:
Ok, how is this?

ltratt commented:
Perfect!

bors r+
